src/openms/source/FORMAT/MzQCFile.cpp

```diff
@@ -0,0 +1,199 @@
+    out["MzQC"]["version"] = "1.0.0";
+
+    // optional: contactName, contactAddress, description
+    if (!contactName.empty()) 
+    {
+      out["MzQC"]["contactName"] = contactName;
+    }
+    if (!contactAddress.empty()) 
+    {
+      out["MzQC"]["contactAddress"] = contactAddress;
+    }
+    if (!description.empty()) 
+    {
+      out["MzQC"]["description"] = description;
+    }
+
+    out["MzQC"]["runQualities"]["metadata"]["label"] = label;
+    out["MzQC"]["runQualities"]["metadata"]["inputFiles"] = {
+        {
+          {"location", File::absolutePath(inputFileName)},
+          {"name", File::basename(inputFileName)},
+          {"fileFormat",
+          {
+            {"accession", "MS:10000584"},
+            {"name", "mzML format"}
+          }
+          }
+        },
+          "fileProperties",
+          {
+            {
+              {"accession", "MS:1000747"},
+              {"name", "completion time"},
+              {"value", String(exp.getDateTime().getDate() + "T" + exp.getDateTime().getTime())}
+            },
+            {
+              {"accession", "MS:1000569"},
+              {"name", "SHA-1"},
+              {"value", String(FileHandler::computeFileHash(inputFileName))}
+            },
+            {
+              {"accession", "MS:1000031"},
+              {"name", "instrument model"},
+              {"value", String(exp.getInstrument().getName())}
+            }
+          }
+    };
+
+    VersionInfo::VersionDetails version = VersionInfo::getVersionStruct();
+    out["MzQC"]["runQualities"]["metadata"]["analysisSoftware"] = { // todo
+        {
+          {"accession", "MS:1009001" }, // create new qc-cv for QCCalculator: MS:1009001 quality control metrics generating software
+          {"name", "QCCalculator"},
+          {"version", String(version.version_major)+"."+String(version.version_minor)+"."+String(version.version_patch)},
+          {"uri", "https://www.openms.de"}
+        }
+    };
+
+    out["MzQC"]["runQualities"]["qualityMetrics"] = {};
+    
+    out["MzQC"]["controlledVocabularies"] = {
+        {
+          {"name", "Proteomics Standards Initiative Quality Control Ontology"},
+          {"uri", "https://github.com/HUPO-PSI/mzQC/blob/master/cv/qc-cv.obo"},
+          {"version", "0.1.2"},
+        },
+        {
+          {"name", "Proteomics Standards Initiative Mass Spectrometry Ontology"},
+          {"uri", "https://github.com/HUPO-PSI/psi-ms-CV/blob/master/psi-ms.obo"},
+          {"version", "4.1.49"}
+        }
+    };
+
+    // precursors QC:0000044 not in cv any more
+
+    map<Size, UInt> counts;
+    for (const auto& spectrum : exp)
+      {
+        const Size level = spectrum.getMSLevel();
+        ++counts[level];  // count MS level
+      }
+
+    // create qualityMetric in json format
+    auto addValue = [](const String& accession, const String& name, const auto& value) -> json
+      {
+        ControlledVocabulary cv;
```

Comment:
only load the ControlledVocabulary once!  maybe directly in line 144
you can capture it in the lambda with [&cv]

Reply:
nice!